emails: Add infrastructure to prevent double-sending of custom emails

wandrew0 opened the pull request:
Followed Tim's comments in Issue #19529 to implement a new event type of CUSTOM_EMAIL_SENT to prevent double-sending of customer emails.

After an email is sent, an event of type CUSTOM_EMAIL_SENT is logged in RealmAuditLog.
RealmAuditLog's (realm, event_type, acting_user_id) is set to (realm, CUSTOM_EMAIL_SENT, recipient's user id).
RealmAuditLog's extra_data is set to {"email_id": {email_id}} where {email_id} is the digest of the email template's content.
Before sending an email, check in RealAuditLog if any row with the same {email_id} and recipient's {user_id} exists; if it exists, we skip sending the email.  

[discussion link](
https://chat.zulip.org/#narrow/stream/3-backend/topic/Add.20CUSTOM_EMAIL_SENT.20event.20to.20prevent.20duplicate.20email)

Fixes: issue #19529

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

wandrew0 commented:
I have the following questions below.

- do we periodically truncate RealAuditLog? I arbitrarily queried the past week's audit logs to decide if any previous email of the same content was sent; I wonder if that's correct.
- I don't understand Tim's following comments very well. "We will want to do this exclusion carefully to ensure that logic like the --marketing option, which is designed only to email an address once even if they have multiple accounts for their email address, "
If user_profile is not present, I tried to find it using the to_email address in the RealmAuditLog creation. I'm not sure if that's correct.

timabbott commented:
`RealmAuditLog` is permanent.

timabbott commented on zerver/lib/send_email.py:
```diff
@@ -514,6 +516,14 @@ def get_header(option: Optional[str], header: Optional[str], name: str) -> str:
+        text = f.read()
+        message = Parser(policy=default).parsestr(text)
+        hash_code = hashlib.sha256(text.encode()).hexdigest()[0:32]
+        return message, hash_code
```
Extracting this function seems like a good refactor, but you should do it in its own preparatory commit, not squashed with your actual changes.

Check out our [commit discipline guide](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for more details.

timabbott replied:
This first commit should have a commit message explaining these changes, not a duplicate of that for the main PR.

Read a dozen random recently merged PRs and you'll see some good examples.

timabbott commented on zerver/lib/send_email.py:
```diff
@@ -562,18 +568,52 @@ def send_one_email(
+            if to_user_id is not None:
+                RealmAuditLog.objects.create(
+                    realm=context["realm"],
+                    event_type=RealmAuditLog.CUSTOMER_EMAIL_SENT,
```
This should be `CUSTOM_EMAIL_SENT`.

wandrew0 committed:
emails: refactor send_email.py to create parse_email_template function.

Fix issue#19529

wandrew0 commented:
refactor of send_email.py

wandrew0 commented on zerver/lib/send_email.py:
```diff
@@ -562,18 +568,52 @@ def send_one_email(
+
     return send_one_email
 
 
```
ok, I will, thanks.

wandrew0 replied:
> The `distinct_email` parameter to `send_custom_email` is what I was referring to -- we need to make sure none of the user accounts with that email address have the `RealmAuditLog.CUSTOM_EMAIL_SENT` entry.
> 
> I think that may mean you need to rearrange the logic, or maybe write the audit log entry for every user account in the `distinct` set. Not sure what is better; can you start a discussion in #backend about this detail, and cross-link the threads?

I have changed my logic and updated unit tests. My code changes are described in [zulip:back-end](https://chat.zulip.org/#narrow/stream/3-backend/topic/Add.20CUSTOM_EMAIL_SENT.20event.20to.20prevent.20duplicate.20email)

alya commented:
@wandrew0 Is this PR ready for another round of review?

wandrew0 commented:
> @wandrew0 Is this PR ready for another round of review?

@alya yes. I also started a discussion on [CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/Add.20CUSTOM_EMAIL_SENT.20event.20to.20prevent.20duplicate.20email) per Tim's suggestion.

alya commented:
@timabbott Ready for you again, unless you'd like someone else to take a look.

wandrew0 commented:
@timabbott Thank you for your feedback. I have fixed the commit messages, and it's ready for review now.

timabbott commented on zerver/lib/send_email.py:
```diff
@@ -601,17 +601,40 @@ def send_custom_email(
+                users_with_same_email[user.id] = [user]
+                continue
+            users_with_same_email[prev_user_id].append(user)
+        users = users.distinct("lower_email")
```
I think this works, but why not use a set or `defaultdict(list)` data structure to gather those with matching email addresses, rather than this `prev_email` based logic? That seems like it'd be cleaner.

wandrew0 replied:
Yes, I have changed send_mail.py. Please see the updated diff. Thank you.

timabbott commented on zerver/tests/test_email_notifications.py:
```diff
@@ -234,6 +236,198 @@ def test_send_custom_email_dry_run(self) -> None:
+                logged_event.extra_data.get("email_template_digest")
+                for logged_event in logged_events
+            }
+            self.assertEqual(hash_code_set, {hash_code})
```
I think these could be better done as end-to-end tests that mainly verify that duplicate emails aren't sent, with checking the state of the audit logs being a side detail. See https://zulip.readthedocs.io/en/latest/testing/philosophy.html

wandrew0 replied:
@timabbott Can you clarify? In tests, I checked the duplicate emails weren't sent  by : 
assert_length(mail.outbox, email_count) with email_count tracks the expected emails sent and mail.outbox tracks the actual emails sent.

wandrew0 committed:
emails: Add infrastructure to prevent double-sending of custom emails.

Add CUSTOM_EMAIL_SENT event to log emails sent to
a delivery address. send_custom_email skips sending a user email if
an event with the same email address and same content exists.

Fix #19529
